### PR TITLE
Prevent ByteToMessageDecoder from overreading when !isAutoRead

### PR DESCRIPTION
**Pull request by iamaleksey**
Motivation:

ByteToMessageDecoder only looks at the last channelRead() in the batch
of channelRead()-s when determining whether or not it should call
ChannelHandlerContext#read() to consume more data when !isAutoRead. This
will lead to read() calls issued unnecessaily and unprompted if the very
last channelRead() didn't result in at least one decoded message, even
if there have been messages decoded from other channelRead()-s in the
current batch.

Modifications:

Track decode outcomes for the entire batch of channelRead() calls and
only issue a read in BTMD if the entire batch of channelRead() calls
yielded no complete messages.

Result:

ByteToMessageDecoder will no longer overread when the very last read
yielded no message, but the batch of reads did.

### CODE REVIEW

**Comment by iamaleksey** [comment]
This PR makes `ByteToMessageDecoder` behave *exactly* like `SslHandler` - which overrides this logic - down to flag name and field comment.

This is important to anyone who doesn't use autoread and relies on 'manually' triggering the reads for any reason - which in our case was the implementation of backpressure/flow control.

**Comment by normanmaurer** [comment]
@iamaleksey could you also add a test case ?

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by iamaleksey** [comment]
@normanmaurer of course 👍 Just didn't want to leave you hanging after that conversation. Probably won't be today, but should make it happen by end of the week.

**Comment by carl-mastrangelo** [comment]
Code generally looks good, but a test is something I would like to see too!

**Comment by iamaleksey** [comment]
A simple test added. Please let me know if you want to see more than this - and I'll probably get to it eventually.

File: codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java [hunk 1]
```diff
@@ -345,4 +347,55 @@ public CompositeByteBuf addComponent(boolean increaseWriterIndex, ByteBuf buffer
...
+
+        // 1 complete frame, 0 partial frames: should NOT trigger a read
+        channel.writeInbound(wrappedBuffer(new byte[] { 14 }));
+        assertEquals(2, interceptor.readsTriggered);
+
+        for (int i = 0; i < 5; i++) {
+            ByteBuf read = channel.readInbound();
+            assertEquals(i * 3 + 0, read.getByte(0));
+            assertEquals(i * 3 + 1, read.getByte(1));
+            assertEquals(i * 3 + 2, read.getByte(2));
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java]: read.release() is missing.

File: codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java [hunk 2]
```diff
@@ -345,4 +347,55 @@ public CompositeByteBuf addComponent(boolean increaseWriterIndex, ByteBuf buffer
             assertEquals(0, in.refCnt());
         }
     }
+
+    @Test
+    public void testDoesNotOverRead() {
+        class ReadInterceptingHandler extends ChannelDuplexHandler {
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java]: `extends ChannelOutboundHandlerAdapter` (as you not intercept any inbound events).

File: codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java (hunk 1 above)
Comment (iamaleksey) [reply codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java]: Not anymore (:

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@carl-mastrangelo @ejona86 PTAL again as now there is a unit test.

**Comment by normanmaurer** [comment]
@iamaleksey thanks a lot!

**Pull request merged**